src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -172,134 +176,133 @@ private unsafe void RegisterToCancelPendingIO(NativeOverlapped* overlapped, Canc
+        /// <param name="overlapped">The overlapped that was used for this operation. Will be freed if the operation result will be handled synchronously.</param>
+        /// <param name="cancellationToken">The cancellation token to use to cancel the operation.</param>
+        /// <returns>The result status of the operation.</returns>
+        private unsafe SocketError ProcessIOCPResultWithDeferredSingleBufferPinning(bool success, int bytesTransferred, NativeOverlapped* overlapped, CancellationToken cancellationToken = default)
+        {
+            Debug.Assert(_asyncProcessingState == AsyncProcessingState.InProcess);
 
-                // Completed synchronously, but the handle wasn't marked as skip completion port on success,
-                // so we still need to fall through and behave as if the IO was pending.
+            SocketError socketError = GetIOCPResult(success, overlapped);
+
+            if (socketError == SocketError.IOPending)
+            {
+                RegisterToCancelPendingIO(overlapped, cancellationToken);
+
+                _singleBufferHandle = _buffer.Pin();
+
+                _asyncProcessingState = AsyncProcessingState.Set;
             }
             else
             {
-                // Get the socket error (which may be IOPending)
-                socketError = SocketPal.GetLastSocketError();
-                if (socketError != SocketError.IOPending)
-                {
-                    // Completed synchronously with a failure.
-                    _singleBufferHandleState = SingleBufferHandleState.None;
-                    FreeNativeOverlapped(overlapped);
-                    FinishOperationSyncFailure(socketError, bytesTransferred, SocketFlags.None);
+                _asyncProcessingState = AsyncProcessingState.None;
+                FinishOperationSync(socketError, bytesTransferred, SocketFlags.None);
+            }
 
-                    if (SocketsTelemetry.Log.IsEnabled() && !_disableTelemetry) AfterConnectAcceptTelemetry();
+            return socketError;
+        }
 
-                    return socketError;
-                }
+        /// <summary>Handles the result of an IOCP operation that supports cancellation.</summary>
+        /// <param name="success">true if the IOCP operation indicated synchronous success; otherwise, false.</param>
+        /// <param name="bytesTransferred">The number of bytes transferred, if the operation completed synchronously and successfully.</param>
+        /// <param name="overlapped">The overlapped that was used for this operation. Will be freed if the operation result will be handled synchronously.</param>
+        /// <param name="cancellationToken">The cancellation token to use to cancel the operation.</param>
+        /// <returns>The result status of the operation.</returns>
+        private unsafe SocketError ProcessIOCPResultWithCancellation(bool success, int bytesTransferred, NativeOverlapped* overlapped, CancellationToken cancellationToken)
```

Comment:
It looks like this method is identical to the one above except that it doesn't do `_singleBufferHandle = _buffer.Pin();`?  There's no way to consolidate them?  Note that `default(Memory<T>).Pin()` should be very cheap.

Reply:
Yeah, you're right. I will consolidate them.

I could either pass default(Memory<T>) or make the param nullable and pass null here.

Reply:
Not sure how would it look by passing `Memory` as an arg, but it can be also a bool arg:

```C#
private unsafe SocketError ProcessIOCPResultWithCancellation(
    bool success, 
    int bytesTransferred,
    NativeOverlapped* overlapped,
    bool pinBuffer, // omit "_singleBufferHandle = _buffer.Pin();" if false
    CancellationToken cancellationToken)
```

Reply:
Not sure how would it look by passing `Memory` as an arg, but it can be also a bool arg:

```C#
private unsafe SocketError ProcessIOCPResultWithCancellation(
    bool success, 
    int bytesTransferred,
    NativeOverlapped* overlapped,
    bool pinBuffer, // omit "_singleBufferHandle = _buffer.Pin();" if false
    CancellationToken cancellationToken)
```

Reply:
Not sure how would it look by passing `Memory` as an arg, but it can be also a bool arg:

```C#
private unsafe SocketError ProcessIOCPResultWithCancellation(
    bool success, 
    int bytesTransferred,
    NativeOverlapped* overlapped,
    bool pinBuffer, // omit "_singleBufferHandle = _buffer.Pin();" if false
    CancellationToken cancellationToken)
```

Reply:
I plan to use this for cancellable Accept in the future, and I think that will work better if the buffer is passed in (it's not always just _buffer in the Accept case).